### ISSUE TITLE
Child States are not waiting for parent state promise resolves 

### ISSUE THREAD

**Issue by samithaf** [angular-ui/ui-router#1903]
Hi Guys,
According to the documentation child states should be able to access parent state resolved promises. In the documentation, https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views#inherited-resolved-dependencies "To provide resolved dependencies via resolve for use by child states." However it seems like when I check the property exists on state change start method, parent state promise has not been resolved. 

I have created a small app based on  http://scotch.io sample application. As you can see in authorize method I am checking user object exists in child state. I am setting the property in parent state. But property has not been set. Please see the sample app in http://plnkr.co/edit/hS8jKONk0ZfyJfE4r7Ty?p=preview

**Comment by eddiemonge** [comment]
The child states need the parent state as a dependency in order for it to wait to load, even if its not used directly.

**Comment by nateabele** [comment]
Perhaps this belongs in the FAQ?

**Comment by BrianMcBrayer** [comment]
Can you give an example of how a child state includes a parent state as a dependency? I cannot inject the parent state or the parent state's resolver into the child state.

**Comment by eddiemonge** [comment]
```coffeescript
angular.module 'example'.config ( $stateProvider ) ->
  $stateProvider.state 'parent',
    resolve: User: ( UserModel ) ->
      UserModel.query().$promise

...

  $stateProvider.state 'parent.child',
    resolve: Profile: ( User ) ->
      Profile.get( User.id ).$promise

```

**Comment by zhdanovartur** [comment]
+1

**Comment by BrianMcBrayer** [comment]
+1. Thanks so much @eddiemonge

**Comment by bmkrocks1** [comment]
+1. I'm having the same issue.

**Comment by eddiemonge** [comment]
@bmkrocks1 which is what? Did you do like I showed in the example?

**Comment by aligajani** [comment]
@eddiemonge , I am having a problem which is similar to this. My child state shows a complete blank page, except the navigation bar, when it is refresh via the browser. My parent state uses a `$http` service to get data from the backend, and I think the child state is executing before the parent state is somehow resolved. If do it step by steps via ui-srefs, it works absolutely fine.

**Comment by mikrich** [comment]
+1

**Comment by geori** [comment]
We are having this same issue with our app.  The app was working with the solution Eddie mentioned above.  I wrote it off as a problem due to Angular 1.5 Beta, but it could be because of a newer version of ui-router (we are on 0.2.15)

**Comment by eddiemonge** [comment]
without code or examples its hard to troubleshoot the problems everyone else is having

**Comment by geori** [comment]
Fortunately for ui-router, I got some code review on my app.  There had been a change to the promise chain and this caused the error.  I was not returning the invocation of a factory function.  See:

```
// parent
AppController.resolve = {
  authUser: ['$state', 'AuthFactory', 'BookmarksFactory'
    function ($state, AuthFactory, BookmarksFactory) {
      return AuthFactory.isAuthenticated()
        .then( function(){ BookmarksFactory.load(); });
  }]
};

// child
ConversationsController.resolve = {
  default: ['$q', 'authUser',
    function ($q, authUser) {
      console.log('resolving new conversation');
      console.log(authUser);
      return $q.when();
    }
  ]
};
```

The problem was occurring because i did not have a return in front of the BookmarksFactory.load() promise: 
```  .then( function(){ BookmarksFactory.load(); }); ```

**Comment by AdrianLungu** [comment]
I'm also having an issue with this with the following case:

```
.state('admin', {
                abstract: true,
                url: "/admin",
                templateUrl: "admin.html",
                resolve: {
                    deps: ['$ocLazyLoad', function ($ocLazyLoad) {
                        return $ocLazyLoad.load({
                            serie: true,
                            files: [
                                'bower_components/ng-tags-input/ng-tags-input.min.css',
                                'bower_components/ng-tags-input/ng-tags-input.bootstrap.min.css',
                                'bower_components/ng-tags-input/ng-tags-input.min.js',
                            ]
                        });
                    }],
                }
            })
.state('admin.super', {
                url: "",
                templateUrl: "super.html",
                controller: "adminSuperController",
                controllerAs: "vm",
                resolve: {
                    deps: ['$ocLazyLoad', function ($ocLazyLoad) {
                        return $ocLazyLoad.load({
                            serie: true,
                            files: [
                                'admin/super/admins.js',
                                'admin/super/service.js',
                            ]
                        });
                    }],
                }
            })
```

And I get the following error:

`Error: [$injector:nomod] Module 'ngTagsInput' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument. http://errors.angularjs.org/1.4.9/$injector/nomod?p0=ngTagsInput at http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:68:12 at http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:2006:17 at ensure (http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:1930:38) at module (http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:2004:14) at L (http://test.takeme.xyz:8888/voxbox/app/bower_components/oclazyload/dist/ocLazyLoad.min.js:8:460) at http://test.takeme.xyz:8888/voxbox/app/bower_components/oclazyload/dist/ocLazyLoad.min.js:8:7109 at processQueue (http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:14991:28) at http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:15007:27 at Scope.$eval (http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:16251:28) at Scope.$digest (http://test.takeme.xyz:8888/voxbox/app/bower_components/angular/angular.js:16069:31)`

Now, I'm not sure if I'm misinterpreting the documentation but from what I've read, the child inherits the resolve from the parent, but in this case, it seems the child is resolving before the parent and thus not being able to display the page because it can't find the dependencies in the parent.

**Comment by eddiemonge** [comment]
Like I said [in my first comment](https://github.com/angular-ui/ui-router/issues/1903#issuecomment-95715971), you have to require the parent in the child.

**Comment by AdrianLungu** [comment]
Oh, don't know how I missed that, it's working now, thanks!

**Comment by Deleugpn** [comment]
@AdrianLungu how did you solve your problem? I can't understand anything happening in the snippet provided by eddiemonge, but your code is a lot more familiar and your solution could help me.

Thanks.

**Comment by AdrianLungu** [comment]
@Deleugpn this is the code I ended up with:

`.state('admin', {
                abstract: true,
                templateUrl: 'admin/index.html',
                url: "/admin",
                resolve: {
                    depsAdmin: ['$ocLazyLoad', function ($ocLazyLoad) {
                        return $ocLazyLoad.load({
                            name: 'app.admin.core',
                            serie: true,
                            files: [
                                'admin/core/core.module.js',
                                'admin/core/core.service.js',

                                'admin/core/directives/orders/orders.js',
                                'admin/core/directives/orders/order.modal.js',
                                'admin/core/directives/admin-management.directive.js',
                                'admin/core/directives/emissionGraph/emission-edit.modal.js',
                                'admin/core/directives/emissionGraph/emission-graph.js',
                                'admin/core/directives/reports/reports.js',
                            ]
                        });
                    }]
                }
            })

.state('admin.super', {
                abstract: true,
                url: "/super",
                templateUrl: "admin/super/index.html",
                data: {
                    permissions: {
                        only: ['superAdmin'],
                        redirectTo: 'admin.login'
                    }
                },
                resolve: {
                    depsAdminSuper: ['$ocLazyLoad', 'depsAdmin', function ($ocLazyLoad, depsAdmin) {
                        return $ocLazyLoad.load({
                            name: 'app.admin.super',
                            serie: true,
                            files: [
                                'admin/super/super.module.js',
                                'admin/super/header.directive.js',
                            ]
                        });
                    }],
                    $title: function () {
                        return 'Super Administration';
                    }
                }
            })

`

The difference is very subtle, but basically it all comes down to these 2 lines:
`                    depsAdmin: ['$ocLazyLoad', function ($ocLazyLoad) {
`
and
`                    depsAdminSuper: ['$ocLazyLoad', 'depsAdmin', function ($ocLazyLoad, depsAdmin) {
`

Basically, named the parent dependency array "depsAdmin" and gave it as a parameter to the child dependency "depsAdminSuper". 

Hope that helps!

**Comment by abrahamD** [comment]
Is there a reason why the child state does not wait for the parent's resolves? It seems to me this should be the default behaviour.

**Comment by amalitsky** [comment]
Seems like it works as expected in v.1.
> you have to require the parent in the child.
Should have been a part of documentation for 0.2.

**Comment by TanaseButcaru** [comment]
Thanks @amalitsky, it worked!

```
'parent':
resolve: 
    parentData: ...

'parent.child':
resolve:
   childData: ['parentData', ...
```

**Comment by christopherthielen** [comment]
> Is there a reason why the child state does not wait for the parent's resolves? It seems to me this should be the default behaviour.

@abrahamD

## 0.x legacy behavior

In 0.x, all resolves are "eager".  If you transition from state `foo` to `parent.child.grandchild`, the resolves for `parent`, `parent.child`, and `parent.grandchild` will start loading eagerly.  **As soon as all of a resolve's dependencies are ready**, the resolve will load.  

If there is an **implicit dependency** between parent and child resolves, the child resolve should inject the parent resolve, so it will wait for the parent to complete before the child starts.

## 1.0 behavior

In 1.0, we added different resolve policies of `LAZY` and `EAGER`.  

In 1.0, resolves by default are "LAZY".  A state's LAZY resolves will wait for the parent state's resolves to finish before beginning to load (in reality, LAZY resolves begin [when the state is being entered](https://github.com/ui-router/core/blob/master/src/hooks/resolve.ts#L27-L43))

We made this change due to feedback like yours.  Each state's resolves waiting for the parent states' seems the least surprising.  If anyone wants "EAGER" behavior they can opt in.

**Comment by abrahamD** [comment]
It's good to see a project evolve with user feedback. Thank you for the response @christopherthielen.

**Comment by matt-webfab** [comment]
Been chasing my tail on this, and I'm glad to find an explanation (lazy/eager), but the response by @christopherthielen leaves me a bit confused.

He says for 0.x, "If there is an *implicit dependency* between parent and child resolves, the child resolve should inject the parent resolve, so it will wait for the parent to complete before the child starts." It seems that if a state is a child of another, that in itself is implicit, and that *explicit* would be if I specifically typed in things to inject.

Also it took me a bit to figure out what was meant by eager and lazy, which I now understand to mean async (just run them and I don't care what order they come back or when) and sync (chained in order as proper dependencies and don't do anything else till they're done). The html script tag "async" attribute behaves this way.

Semantics aside, using 0.x requires me to add a dummy resolver with parent dependencies for every child state, which is not "dry", and exactly what I was trying to avoid by using child states to begin with ("globally" waiting for firebase to initialize and checking login status before deciding if a user has privileges to go where they were trying to, in both "hot" and "cold" situations). I'm pasting these dummies on everything and getting odd results. Switching versions mid-project isn't an option. Is there a magic config I can toggle?

**Comment by brphelps** [comment]
No magic config, although my team is creating something which is "magically" wrapping child states with this logic to emulate what we think should be expected behavior.

**Comment by christopherthielen** [comment]
@matt-webfab 

> Also it took me a bit to figure out what was meant by eager and lazy

to clarify,

- Eager: all eager resolves [begin loading when the transition STARTS via an `onStart` hook](https://github.com/ui-router/core/blob/afdeec286b68a5df5d74b7093f682354c40e26ef/src/hooks/resolve.ts#L11-L26).  The transition waits for the resolves to complete before continuing.
- Lazy: all lazy resolves for a state [begin loading when the state is being entered via an `onEnter` hook](https://github.com/ui-router/core/blob/afdeec286b68a5df5d74b7093f682354c40e26ef/src/hooks/resolve.ts#L28-L44).  The transition waits for the resolves before the state is done being entered.

> using 0.x requires me to add a dummy resolver with parent dependencies for every child state, which is not "dry", and exactly what I was trying to avoid by using child states to begin with

No, in 0.x *all the resolves are processed first* (eagerly).  When they are done, then states are entered.  in 0.x, resolves are processed *as soon as possible* (eagerly), when all their dependencies are ready. 

Adding a dummy resolver with parent dependencies to a state *has no effect on the state being entered*.  Adding dummy parent dependencies to a resolve does have an effect.

In 0.x, the only way to implement what you are attempting is with a state change start event (cancel the transition, ensure the user is authenticated, then re-run the transition), or via [deferIntercept](https://ui-router.github.io/ng1/docs/0.4.2/#/api/ui.router.router.$urlRouterProvider) to when the application is first started.  The problems with doing these things is precisely why ui-router 1.0 abandoned the state change events in favor of Transition Hooks, which allow you to implement your use case easily.

**Comment by nermand** [comment]
If you have multiple promises being resolved in the child state, only those depending on the parent's promise will wait for it to be resolved. 
So, it's not like **all** resolutions wait for the parent, rather they are run as async, except the one actually being injected with parent's promise.

I had problems making it work, so I hope this comment will save time for someone else.
